### PR TITLE
fix: Improve continue button on org-add-team and validations [CAL-2015]

### LINKED ISSUE #10112: Disable button that creates a Team till the user writes something
### Issue Summary

This is not really a bug but an enhancement regarding UI/UX.
During the organization creation flow, it asks to `Create your Teams` step. Acting by impulse, the user could click on the `Continue` button. This will show a notification saying that a team name is mandatory. It could be good to have that button disabled till the user write something in the blank field. 

Note: Maybe a placeholder would be good as well. Or, something that indicates this is a required field.

image

This can be the same for the Organizations admin. We can skip that, I think is the same situation.
image

The last thing, it would be good to have consistency regarding UI/UX when a field is not filled out. From organization one, this is a self-component notification. From Team, it is using another component, like a Toast.

image

image

### PR DESCRIPTION
**Pull request by alannnc**
## What does this PR do?

- disables button if no team name it's entered
- validates all inputs have team names and no duplicates
- Fixes some validations on the server before creating new team

Fixes #10112 

https://www.loom.com/share/c5ed930f84924e569d4f00a2c1d9fad8

## Requirement/Documentation

  - If there is a requirement document, please, share it here.
  - If there is ab UI/UX design document, please, share it here.

## Type of change

  - [X] Bug fix (non-breaking change which fixes an issue)

## How should this be tested?

  - Are there environment variables that should be set? No
  - What are the minimal test data to have? A newly created org team on the onboarding page step 5
  - What is expected (happy path) to have (input and output)? Any team name should be valid
  - Any other important info that could help to test that PR. No

## Mandatory Tasks

  - [ ] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

## Checklist

- I haven't added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: packages/features/ee/organizations/components/AddNewTeamsForm.tsx [hunk 1]
```diff
@@ -1,12 +1,11 @@
-import { ArrowRight } from "lucide-react";
 import { useRouter } from "next/router";
 import { useState } from "react";
 import { z } from "zod";
 
 import { useLocale } from "@calcom/lib/hooks/useLocale";
 import { trpc } from "@calcom/trpc/react";
 import { Button, showToast, TextField } from "@calcom/ui";
-import { Plus, X } from "@calcom/ui/components/icon";
+import { Plus, X, ArrowRight } from "@calcom/ui/components/icon";
```
Comment (alannnc) [inline packages/features/ee/organizations/components/AddNewTeamsForm.tsx:8]: Can be imported from calcom lib

File: packages/features/ee/organizations/components/AddNewTeamsForm.tsx [hunk 2]
```diff
@@ -18,7 +17,7 @@ export const AddNewTeamsForm = () => {
   const { id: orgId } = querySchema.parse(router.query);
   const [counter, setCounter] = useState(1);
 
-  const [inputValues, setInputValues] = useState<string[]>([""]);
+  const [inputValues, setInputValues] = useState<string[]>([]);
```
Comment (alannnc) [inline packages/features/ee/organizations/components/AddNewTeamsForm.tsx:20]: Can be initialized as empty array

File: packages/features/ee/organizations/components/AddNewTeamsForm.tsx [hunk 3]
```diff
@@ -42,6 +41,8 @@ export const AddNewTeamsForm = () => {
     async onSuccess(data) {
       if (data.duplicatedSlugs.length) {
         showToast(t("duplicated_slugs_warning", { slugs: data.duplicatedSlugs.join(", ") }), "warning");
+        // Server will return array of duplicated slugs, so we need to wait for user to read the warning
+        // before pushing to next page
```
Comment (alannnc) [inline packages/features/ee/organizations/components/AddNewTeamsForm.tsx:45]: Comment to help understand the flow of submit

File: packages/features/ee/organizations/components/AddNewTeamsForm.tsx [hunk 4]
```diff
@@ -54,21 +55,42 @@ export const AddNewTeamsForm = () => {
...
+      return false;
+    }
+    return true;
+  };
+
+  const handleFormSubmit = () => {
+    if (validateInput()) {
+      createTeamsMutation.mutate({ orgId, teamNames: inputValues });
+    }
+  };
```
Comment (alannnc) [inline packages/features/ee/organizations/components/AddNewTeamsForm.tsx:75]: Frontend validations

File: packages/features/ee/organizations/components/AddNewTeamsForm.tsx [hunk 5]
```diff
@@ -54,21 +55,42 @@ export const AddNewTeamsForm = () => {
...
+
   return (
     <>
       {Array.from({ length: counter }, (_, index) => (
         <div className="relative" key={index}>
           <TextField
             key={index}
-            value={inputValues[index]}
+            value={inputValues[index] || ""}
```
Comment (alannnc) [inline packages/features/ee/organizations/components/AddNewTeamsForm.tsx:83]: As inputValues can be undefined we set value on textField to empty string so react doesn't complain about
controlled/uncontrolled inputs.

File: packages/features/ee/organizations/components/AddNewTeamsForm.tsx [hunk 6]
```diff
@@ -54,21 +55,42 @@ export const AddNewTeamsForm = () => {
...
             onChange={(e) => handleInputChange(index, e.target.value)}
             addOnClassname="bg-transparent p-0 border-l-0"
+            placeholder={index === 0 ? t("org_team_names_example") : ""}
```
Comment (alannnc) [inline packages/features/ee/organizations/components/AddNewTeamsForm.tsx:86]: Only displaying placeholder on first input

File: packages/features/ee/organizations/components/AddNewTeamsForm.tsx [hunk 7]
```diff
@@ -54,21 +55,42 @@ export const AddNewTeamsForm = () => {
...
             addOnSuffix={
               index > 0 && (
                 <Button
                   color="minimal"
                   className="group/remove mx-2 px-0 hover:bg-transparent"
-                  onClick={() => handleRemoveInput(index)}>
+                  onClick={() => handleRemoveInput(index)}
+                  aria-label="Remove Team">
```
Comment (alannnc) [inline packages/features/ee/organizations/components/AddNewTeamsForm.tsx:93]: should change to translation string

File: packages/features/ee/organizations/components/AddNewTeamsForm.tsx [hunk 8]
```diff
@@ -80,26 +102,16 @@ export const AddNewTeamsForm = () => {
...
-            showToast(t("team_names_empty"), "error");
-          } else {
-            const duplicates = inputValues.filter((item, index) => inputValues.indexOf(item) !== index);
-            if (duplicates.length) {
-              showToast("team_names_repeated", "error");
-            } else {
-              createTeamsMutation.mutate({ orgId, teamNames: inputValues });
-            }
-          }
-        }}>
```
Comment (alannnc) [inline packages/features/ee/organizations/components/AddNewTeamsForm.tsx:102]: Moved function to own declaration

File: packages/trpc/server/routers/viewer/organizations/createTeams.handler.ts [hunk 9]
```diff
@@ -16,31 +16,57 @@ type CreateTeamsOptions = {
...
+  const user = ctx.user;
 
-  const organization = await prisma.team.findFirst({
-    where: { id: orgId },
-    select: { slug: true, metadata: true },
-  });
-  const metadata = teamMetadataSchema.parse(organization?.metadata);
+  if (!user) {
+    throw new NoUserError();
+  }
```
Comment (alannnc) [inline packages/trpc/server/routers/viewer/organizations/createTeams.handler.ts:24]: return early

File: packages/trpc/server/routers/viewer/organizations/createTeams.handler.ts [hunk 10]
```diff
@@ -16,31 +16,57 @@ type CreateTeamsOptions = {
...
+
+  const { teamNames, orgId } = input;
 
-  if (!metadata?.requestedSlug && !organization?.slug)
-    throw new TRPCError({ code: "BAD_REQUEST", message: "no_organization_slug" });
+  if (orgId !== user.organizationId) {
```
Comment (alannnc) [inline packages/trpc/server/routers/viewer/organizations/createTeams.handler.ts:28]: we using orgId from input so we should validate input.orgId matches with user in ctx

File: packages/trpc/server/routers/viewer/organizations/createTeams.handler.ts [hunk 11]
```diff
@@ -16,31 +16,57 @@ type CreateTeamsOptions = {
...
+  const userMembershipRole = await prisma.membership.findFirst({
     where: {
-      userId: ctx.user.id,
+      userId: user.id,
       teamId: orgId,
+      role: {
+        in: ["OWNER", "ADMIN"],
+      },
+      // @TODO: not sure if this already setup earlier
+      // accepted: true,
```
Comment (alannnc) [inline packages/trpc/server/routers/viewer/organizations/createTeams.handler.ts:41]: Validating team creation to specific membership.role

File: packages/trpc/server/routers/viewer/organizations/createTeams.handler.ts [hunk 12]
```diff
@@ -16,31 +16,57 @@ type CreateTeamsOptions = {
...
+  }
+
+  const organization = await prisma.team.findFirst({
+    where: { id: orgId },
+    select: { slug: true, metadata: true },
+  });
+
+  if (!organization) throw new NoOrganizationError();
+
+  const parseTeams = teamMetadataSchema.safeParse(organization?.metadata);
```
Comment (alannnc) [inline packages/trpc/server/routers/viewer/organizations/createTeams.handler.ts:59]: safeParse so we can throw custom error and not 500

File: packages/trpc/server/routers/viewer/organizations/createTeams.handler.ts [hunk 13]
```diff
@@ -51,11 +77,16 @@ export const createTeamsHandler = async ({ ctx, input }: CreateTeamsOptions) =>
     .flatMap((ts) => ts.slug ?? [])
     .concat(userSlugs.flatMap((us) => us.username ?? []));
 
-  const duplicatedSlugs = existingSlugs.filter((slug) => teamNames.includes(slug));
+  const duplicatedSlugs = existingSlugs.filter((slug) =>
+    teamNames.map((item) => slugify(item)).includes(slug)
```
Comment (alannnc) [inline packages/trpc/server/routers/viewer/organizations/createTeams.handler.ts:81]: was validating ["Teams"] with ["teams'] so this validation wasn't case sensitive

File: packages/trpc/server/routers/viewer/organizations/createTeams.handler.ts [hunk 14]
```diff
@@ -51,11 +77,16 @@ export const createTeamsHandler = async ({ ctx, input }: CreateTeamsOptions) =>
...
+  );
 
+  if (duplicatedSlugs.length === teamNames.length) {
+    return { duplicatedSlugs };
+  }
```
Comment (alannnc) [inline packages/trpc/server/routers/viewer/organizations/createTeams.handler.ts:86]: When input teamNames has the same length that deplicatedSlugs that means every input it's duplicated so we should return duplicatedSlugs w/o creating any team.

File: packages/trpc/server/routers/viewer/organizations/createTeams.handler.ts [hunk 15]
```diff
@@ -51,11 +77,16 @@ export const createTeamsHandler = async ({ ctx, input }: CreateTeamsOptions) =>
...
   await prisma.$transaction(
     teamNames.flatMap((name) => {
-      if (!duplicatedSlugs.includes(name)) {
+      if (!duplicatedSlugs.includes(slugify(name))) {
```
Comment (alannnc) [inline packages/trpc/server/routers/viewer/organizations/createTeams.handler.ts:89]: Case sensitive matching

File: packages/trpc/server/routers/viewer/organizations/createTeams.handler.ts [hunk 16]
```diff
@@ -74,3 +105,33 @@ export const createTeamsHandler = async ({ ctx, input }: CreateTeamsOptions) =>
...
+  constructor() {
+    super({ code: "BAD_REQUEST", message: "no_organization_found" });
+  }
+}
+
+class NoOrganizationSlugError extends TRPCError {
+  constructor() {
+    super({ code: "BAD_REQUEST", message: "no_organization_slug" });
+  }
+}
```
Comment (alannnc) [inline packages/trpc/server/routers/viewer/organizations/createTeams.handler.ts:137]: Custom Handler Errors

File: packages/ui/components/form/inputs/Input.tsx [hunk 17]
```diff
@@ -12,7 +12,7 @@ import { Eye, EyeOff, X, Search } from "../../icon";
 import { HintsOrErrors } from "./HintOrErrors";
 import { Label } from "./Label";
 
-type InputProps = JSX.IntrinsicElements["input"] & { isFullWidth?: boolean; isStandaloneField?: boolean };
+type InputProps = JSX.IntrinsicElements["input"] & { isFullWidth?: boolean };
```
Comment (alannnc) [inline packages/ui/components/form/inputs/Input.tsx:15]: This an extra, was receiving error on console, it looks like it doesn't have any use.

File: packages/ui/components/form/inputs/Input.tsx [hunk 18]
```diff
@@ -140,7 +140,6 @@ export const InputField = forwardRef<HTMLInputElement, InputFieldProps>(function
             type={type}
             placeholder={placeholder}
             isFullWidth={inputIsFullWidth}
-            isStandaloneField={false}
```
Comment (alannnc) [inline packages/ui/components/form/inputs/Input.tsx:143]: declared as false an optional on type, no other use found in entire codebase.

**Review by ericrommel — CHANGES_REQUESTED** [review]
Hey @alannnc, almost there... I found some good things to change. If you delete the text written, turning it to empty again, the button should go back to disable status.

https://www.loom.com/share/9edd483ae2dd47a19097b6a147256f24

**Note 1**: If you add a new team, without filling out anything, the button is turned to enabled. If there is no text, the button should remain disabled. In case of at least one field is filled out, then the button should be turned to enable but a notification should be shown in case any other field is empty.
**Note 2**: Would it be good to have placeholders for new fields?

**Comment by alannnc** [comment]
> **Note 1**: If you add a new team, without filling out anything, the button is turned to enabled. If there is no text, the button should remain disabled. In case of at least one field is filled out, then the button should be turned to enable but a notification should be shown in case any other field is empty. **Note 2**: Would it be good to have placeholders for new fields?
Note 1: I leaved it like this because it felt more adequate. But sure I can make this change.
Note 2: To me it felt noisy having the same placeholder on everything. What we can do here it's having probably different team names for every input, having 10 and then repeat. 
Also now that I'm thinking we should put a limit on this form. Does 10 teams sounds reasonable? You can still add them later in other page.
@ericrommel

**Comment by ericrommel** [comment]
> > **Note 1**: If you add a new team, without filling out anything, the button is turned to enabled. If there is no text, the button should remain disabled. In case of at least one field is filled out, then the button should be turned to enable but a notification should be shown in case any other field is empty. **Note 2**: Would it be good to have placeholders for new fields?
> 
> Note 1: I leaved it like this because it felt more adequate. But sure I can make this change. Note 2: To me it felt noisy having the same placeholder on everything. What we can do here it's having probably different team names for every input, having 10 and then repeat. Also now that I'm thinking we should put a limit on this form. Does 10 teams sounds reasonable? You can still add them later in other page. @ericrommel

About Note 1, we end up to back to the original problem. That's why IMAO we should have better control here.
About Note 2, I'm okay with not having placeholders for additional team fields.

Definitely, we should limit. I would go with 5 instead of 10 and have a notification to inform that more can be added later. Good catch ;)

image

**Comment by alannnc** [comment]
@ericrommel Just updated with latest suggestions

**Comment by ericrommel** [comment]
> @ericrommel Just updated with latest suggestions

Hey @alannnc,

Sorry about that but I'm still finding issues. Please, check the video to better understand:

https://www.loom.com/share/1b689d304b844dcfb258f37d89f710ec

**Comment by alannnc** [comment]
> > @ericrommel Just updated with latest suggestions
> 
> Hey @alannnc,
> 
> Sorry about that but I'm still finding issues. Please, check the video to better understand:
> 
> https://www.loom.com/share/1b689d304b844dcfb258f37d89f710ec

For the change, I opted for a secure way, if you added a new team input Continue button goes back to disabled till you fill a value there. Or does it makes more sense allowing input blanks but ignoring them?
@ericrommel

**Comment by ericrommel** [comment]
> > > @ericrommel Just updated with latest suggestions
> > 
> > 
> > Hey @alannnc,
> > Sorry about that but I'm still finding issues. Please, check the video to better understand:
> > https://www.loom.com/share/1b689d304b844dcfb258f37d89f710ec
> 
> For the change, I opted for a secure way, if you added a new team input Continue button goes back to disabled till you fill a value there. Or does it makes more sense allowing input blanks but ignoring them? @ericrommel

@alannnc, you have opted very correctly but the issue is, if you check the Loom if I add all possible fields (5 in our new case) when I fill out the third one (2 are still missing), the button goes to be enabled. If the Loom is not good enough, we can connect and I can show you step by step.

**Comment by alannnc** [comment]
> > > > @ericrommel Just updated with latest suggestions
> > > 
> > > 
> > > Hey @alannnc,
> > > Sorry about that but I'm still finding issues. Please, check the video to better understand:
> > > https://www.loom.com/share/1b689d304b844dcfb258f37d89f710ec
> > 
> > 
> > For the change, I opted for a secure way, if you added a new team input Continue button goes back to disabled till you fill a value there. Or does it makes more sense allowing input blanks but ignoring them? @ericrommel
> 
> @alannnc, you have opted very correctly but the issue is, if you check the Loom if I add all possible fields (5 in our new case) when I fill out the third one (2 are still missing), the button goes to be enabled. If the Loom is not good enough, we can connect and I can show you step by step.

Ahh alright, I watched it several times and got confused. But yeah now that you pointed it I can try to replicate it. Thanks a lot @ericrommel.

**Comment by ericrommel** [comment]
> > > > > @ericrommel Just updated with latest suggestions
> > > > 
> > > > 
> > > > Hey @alannnc,
> > > > Sorry about that but I'm still finding issues. Please, check the video to better understand:
> > > > https://www.loom.com/share/1b689d304b844dcfb258f37d89f710ec
> > > 
> > > 
> > > For the change, I opted for a secure way, if you added a new team input Continue button goes back to disabled till you fill a value there. Or does it makes more sense allowing input blanks but ignoring them? @ericrommel
> > 
> > 
> > @alannnc, you have opted very correctly but the issue is, if you check the Loom if I add all possible fields (5 in our new case) when I fill out the third one (2 are still missing), the button goes to be enabled. If the Loom is not good enough, we can connect and I can show you step by step.
> 
> Ahh alright, I watched it several times and got confused. But yeah now that you pointed it I can try to replicate it. Thanks a lot @ericrommel.

Sorry if I couldn't be clear enough. ;)

**Comment by alannnc** [comment]
> Sorry if I couldn't be clear enough. ;)

Don't worry! I found the issue, was missing some initial state so was validating against -1 input every time thats why 4 input filled worked there.

File: packages/features/ee/organizations/components/AddNewTeamsForm.tsx [hunk 19]
```diff
@@ -54,52 +58,71 @@ export const AddNewTeamsForm = () => {
...
+    }
+    return true;
+  };
+
+  const handleFormSubmit = () => {
+    if (validateInput()) {
+      createTeamsMutation.mutate({ orgId, teamNames: inputValues });
+    }
+  };
+  console.log({ inputValues });
```
Comment (zomars) [inline packages/features/ee/organizations/components/AddNewTeamsForm.tsx:82]:
```suggestion
```

**Review by zomars — CHANGES_REQUESTED** [review]
Overall code LGTM. But app store tests are failing. Will approve once checks and @ericrommel feedback is addressed. 🙏🏽

**Review by zomars — DISMISSED** [review]
Overall code LGTM. But app store tests are failing. Will approve once checks and @ericrommel feedback is addressed. 🙏🏽

**Comment by alannnc** [comment]
@ericrommel I think every issue has been handled already and test are passing now. Can we get an approve for merge?

**Comment by ericrommel** [comment]
> @ericrommel I think every issue has been handled already and test are passing now. Can we get an approve for merge?

@alannnc LGTM. Everything worked fine.

**Review by ericrommel — APPROVED** [review]
@alannnc @zomars the findings for `org-add-team` were fixed. From my side, this is okay.

**Pull request merged**